ratbagd/ratbagd.c

```diff
@@ -204,35 +204,98 @@ static const sd_bus_vtable ratbagd_vtable[] = {
 	SD_BUS_VTABLE_END,
 };
 
+/* 32 pre-picked names. If you have more than 32 devices connected and need
+ * ratbag to control them all, congratulations, you are official a niche
+ * use-case.
+ */
+static const char *device_ids[] = {
+"mara", "capybara", "porcupine", "paca",
+"vole", "woodrat", "gerbil", "shrew",
+"hare", "beaver", "gopher", "chinchilla",
+"mouse", "rat", "rabbit", "guineapig",
+"hutia", "degu", "gundi", "acouchy",
+"nutria", "tuco-tuco", "paca", "hamster",
+"zokor", "chipmunk", "squirrel", "marmot",
+"groundhog", "suslik", "agouti", "blesmol",
+};
+
+/* bitmask telling us which ones of the above are free. */
+static uint32_t device_ids_free = (1UL << ARRAY_LENGTH(device_ids)) - 1;
+
+static inline const char *
+pick_device_id(const char *sysname)
+{
+	int index;
+
+	/* search for the first number in the sysname and where we find it,
+	 * use that as lookup index. This way we get some sort of
+	 * consistency, hidraw2 is always "rat", etc.
+	 * Totally not necessary but we might as well have some fun here.
+	 * Where this fails or picks an id already chosen, we fall back to
+	 * "choose the first one free".
+	 */
+	if (sscanf(sysname, "%*[^0-9]%d", &index) == 1 &&
+	    index >= 0 && (unsigned)index < sizeof(device_ids_free) * 8 &&
+	    !!(device_ids_free & (1 << index)))
+	    goto assign;
+
+	index = ffs(device_ids_free);
```

Comment:
Never knew there was a *for fuck's sake* function in the standard library, interesting!

Reply:
```
$> man ffs

FFS(3)    Linux Programmer's Manual    FFS(3)

NAME
       ffs, ffsl, ffsll - find first bit set in a word

SYNOPSIS
       #include <strings.h>

       int ffs(int i);

       #include <string.h>

       int ffsl(long int i);

       int ffsll(long long int i);

...
```

standard C function, cherry picked by Python :)

Reply:
i can never unsee the real meaning of ffs either :)